### ISSUE TITLE
Startup errors

### ISSUE THREAD

**Issue by datsme** [CitizensDev/Citizens2#1630]
Constant spam errors on startup. I turned on both debug modes in the config then started up the server for the log.    

Link to server log - https://pastebin.com/TVs2H6bg
Citizens config - https://pastebin.com/k1kAprXt

The output of command `/version` on my server is: 
CraftBukkit version git-Spigot-e769fe4-068dab5 (MC: 1.13.2) (Implementing API version 1.13.2-R0.1-SNAPSHOT)
The output of command `/version citizens` on my server is: Citizens v2.0.24-SNAPSHOT (build 1599)

**Comment by mcmonkey4eva** [comment]
Can you post your `Citizens/saves.yml` file? In particular the section for NPC ID 4.

**Comment by datsme** [comment]
Sure. I posted just what you requested below. However I also created a pastebin for the entire file just in case. https://pastebin.com/dJuRUGFT

  '4':
    name: Pastor Jeff
    uuid: 8fa13e8b-399e-4cf1-b74b-566a34e14112
    traits:
      waypoints:
        linear:
          points:
            '0':
              location:
                world: Cassian
                x: '405.0'
                y: '62.0'
                z: '224.0'
                yaw: '0.0'
                pitch: '0.0'
            '1':
              location:
                world: Cassian
                x: '412.0'
                y: '62.0'
                z: '224.0'
                yaw: '0.0'
                pitch: '0.0'
            '2':
              location:
                world: Cassian
                x: '408.0'
                y: '62.0'
                z: '224.0'
                yaw: '0.0'
                pitch: '0.0'
              triggers:
                '0':
                  delay: 60
                  type: delay
        provider: linear
      text:
        talk-close: true
        random-talker: true
        realistic-looking: false
        range: '10.0'
        talkitem: default
        delay: -1
        text:
          '0': For the wages of sin is death, but the gift of God is eternal life
            in Christ Jesus our Lord. Romans 6:23
          '1': for ALL have sinned and fall short of the glory of God Romans 3:23
          '2': They replied, "Believe in the Lord Jesus, and you will be saved - you
            and your household." Acts 16:31
      speech: chat
      lookclose:
        enabled: true
        range: '5.0'
        realisticlooking: false
      type: PLAYER
      gravity:
        enabled: false
      spawned: true
      location:
        world: Cassian
        x: '408.8119'
        y: '63.0'
        z: '224.4932'
        yaw: '88.76832580566406'
        pitch: '0.0'
      owner:
        owner: Dantarith
        uuid: bc8fb122-0955-4c4a-8626-630ed2426d90
    traitnames: text,spawned,waypoints,lookclose,equipment,gravity,type,speech,mounttrait,inventory,poses,location,owner
    navigator:
      speedmodifier: '1.0'
      avoidwater: false
      usedefaultstuckaction: true

Complete File

**Comment by mcmonkey4eva** [comment]
Alright, this looks to be a bug in Citizens internal code we've got to fix.

A bit weird to trace but I had a hunch looking at the save data and I believe I've solidly traced how that ends up breaking.

I'm going to post a runthrough of how it breaks below for reference when going in to fix it.

-----

So:
You have an NPC (id 4) with 3 waypoints.
You have a `delay` trigger on waypoint `0` (the first in set).

When the NPC finishes it's last point in the path, it reaches this code: https://github.com/CitizensDev/Citizens2/blob/62d264bf312eef2d4e2b22b3b5a333c0b0b2839c/main/src/main/java/net/citizensnpcs/npc/ai/CitizensNavigator.java#L174
Which goes down to here: https://github.com/CitizensDev/Citizens2/blob/62d264bf312eef2d4e2b22b3b5a333c0b0b2839c/main/src/main/java/net/citizensnpcs/npc/ai/CitizensNavigator.java#L292-L294
Which throws an error saying the set was modified outside of the execution while it was running.
I believe that's because there is an `onCompletion` call for the delay trigger at this point

That is, this code calls the delay trigger method: https://github.com/CitizensDev/Citizens2/blob/62d264bf312eef2d4e2b22b3b5a333c0b0b2839c/main/src/main/java/net/citizensnpcs/trait/waypoint/Waypoint.java#L96-L105
Which is added as a callback here: https://github.com/CitizensDev/Citizens2/blob/62d264bf312eef2d4e2b22b3b5a333c0b0b2839c/main/src/main/java/net/citizensnpcs/trait/waypoint/LinearWaypointProvider.java#L472-L480

Which ends up with this line being called: https://github.com/CitizensDev/Citizens2/blob/62d264bf312eef2d4e2b22b3b5a333c0b0b2839c/main/src/main/java/net/citizensnpcs/trait/waypoint/triggers/DelayTrigger.java#L40
Which calls this: https://github.com/CitizensDev/Citizens2/blob/62d264bf312eef2d4e2b22b3b5a333c0b0b2839c/main/src/main/java/net/citizensnpcs/trait/waypoint/LinearWaypointProvider.java#L443-L450
Which includes as part of its logic this: https://github.com/CitizensDev/Citizens2/blob/62d264bf312eef2d4e2b22b3b5a333c0b0b2839c/main/src/main/java/net/citizensnpcs/npc/ai/CitizensNavigator.java#L61-L64
Which calls this: https://github.com/CitizensDev/Citizens2/blob/62d264bf312eef2d4e2b22b3b5a333c0b0b2839c/main/src/main/java/net/citizensnpcs/npc/ai/CitizensNavigator.java#L292-L295
... the exact same code, operating on the same list, clearing it out.
When the callstack jumps back up to the original point it's called, it's now been cleared out, and the loop detects concurrent modification and faults.

Long story short: when a delay trigger is present on waypoint 0 of a path, the method `CitizensNavigator#stopNavigating(CancelReason)` indirectly calls itself and breaks its own state.

I'm not sure exactly the ideal fix, but that's I believe the issue trace in full.

@fullwall ^

**Comment by datsme** [comment]
Thanks for that explanation! If it helps this same npc and file worked fine on the 1.12.2 version. It never had an error and he walked the path just fine.
 Anyway, I guess for now I will remove his pathing and just let him stand at his pulpit and be a boring preacher lol ;)  
Thanks again for the support!

**Comment by mcmonkey4eva** [comment]
Should be fixed in Citizens build DEV-1603+ from http://ci.citizensnpcs.co/job/Citizens2/

**Comment by datsme** [comment]
Yes it is fixed. Thank you so much!
